File: prep/implementation/patterns_reference_implementation/sliding_window/NoRepeatSubstring.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class NoRepeatSubstring {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();
      in.nextLine();

      while (tt-- > 0) {
        String s = in.nextLine().trim();

        Map<Character, Integer> pos = new HashMap<>();
        int windowStart = 0, maxLen = 0;

        for (int windowEnd = 0; windowEnd < s.length(); windowEnd++) {
          char r = s.charAt(windowEnd);

          if (pos.containsKey(r)) {
            windowStart = Math.max(windowStart, pos.get(r) + 1);
          }

          pos.put(r, windowEnd);
          maxLen = Math.max(maxLen, windowEnd - windowStart + 1);
        }

        System.out.println(maxLen);
      }
    }
  }
}

File: prep/implementation/patterns_reference_implementation/sliding_window/no_repeat_substring.cpp
#include <algorithm>
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt;
  string s;

  cin >> tt;
  while (tt--) {
    cin >> s;

    int n = s.size();
    int max_len = 0, window_start = 0;
    unordered_map<char, int> pos;

    for (int window_end = 0; window_end < n; window_end++) {
      char c = s[window_end];

      if (pos.find(c) != pos.end()) {
        window_start = max(window_start, pos[c] + 1);
      }
      pos[c] = window_end;
      max_len = max(max_len, window_end - window_start + 1);
    }

    cout << max_len << "\n";
  }

  return 0;
}